src/odemis/acq/stitching/test/tiledacq_test.py

```diff
@@ -311,6 +311,119 @@ def test_whole_procedure(self):
         self.assertIsInstance(data[0], model.DataArray)
         self.assertEqual(len(data[0].shape), 2)
 
+    def test_refocus(self):
+        """Test the refocus function which provides the z levels for the zstack."""
+        area = (-0.001, -0.001, 0.001, 0.001)
+        overlap = 0.2
+        focus_points = [[-0.00019889999999999998, -0.0001485167330631956, 2.5185183e-05],
+                        [0.0, -0.0001485167330631956, -1.8148142e-05],
+                        [0.00019889999999999998, -0.0001485167330631956, -3.382517e-06],
+                        [-0.00019889999999999998, 0.0, 1.3804983e-05],
+                        [0.0, 0.0, 2.320607e-06], [0.00019889999999999998, 0.0, 3e-05],
+                        [-0.00019889999999999998, 0.0001485167330631956, -1.78125e-05],
+                        [0.0, 0.0001485167330631956, -8.124999e-06],
+                        [0.00019889999999999998, 0.0001485167330631956, 4.765627e-06]]
+        focus_range = self.focus.getMetadata().get(model.MD_POS_ACTIVE_RANGE)
+        axis_range = self.focus.axes['z'].range
+
+        # Test 1 when z level is not provided
+        tiled_acq_task = TiledAcquisitionTask(self.fm_streams, self.stage,
+                                              area=area, overlap=overlap, future=model.InstantaneousFuture(),
+                                              focus_points=focus_points)
+        # TODO need the following metadata in mimas-sim and meteor-sim
+        # Only the enzel file is modified to run this testcase that uses enzel-sim
+        # TODO check the values of the metaadat POS_ACTIVE_RANGE in enzel-sim for focus role
+        tiled_acq_task._refocus()
+        # min with none values
+        zmin = max(min(axis_range), min(focus_range))
+        zmax = min(max(axis_range), max(focus_range))
+        # Test the min and max range of the zstack
+        self.assertGreaterEqual(tiled_acq_task._zlevels[0], zmin)
+        self.assertLessEqual(tiled_acq_task._zlevels[-1], zmax)
+
+        # Test 2 when z level is provided
+        zlevels = [focus_points[0][2], focus_points[1][2], focus_points[2][2]]
+
+        with self.assertRaises(NotImplementedError):
+            tiled_acq_task = TiledAcquisitionTask(self.fm_streams, self.stage,
+                                                  area=area, overlap=overlap, future=model.InstantaneousFuture(),
+                                                  zlevels=zlevels, focus_points=focus_points)
```

Comment:
we actually need `zlevels` and `focus_points` to be defined for the `refocus` method to work. The `NotImplementedError` gets raised because you do not set the focusing_method to FocusingMethod.MAX_INTENSITY_PROJECTION. So maybe the code on line 161 of _tiledacq.py should have an additional check:

```
if len(self._zlevels) > 1 and focusing_method != FocusingMethod.MAX_INTENSITY_PROJECTION:
    raise NotImplementedError("Multiple zlevels currently only works with focusing method MAX_INTENSITY_PROJECTION")
elif len(self._zlevels) < 1 and focus_points is not None:
    raise  NotImplementedError("If focus_points are used initial zlevels must also be provided")

```

Reply:
This comment is not correct, sorry. We can also do refocus when the zlevels are not defined. Then it just refocuses the single point.

But I think there should also be a test case that checks with zlevels and focus points and focusing_method = FocusingMethod.MAX_INTENSITY_PROJECTION

Reply:
Just a tiny comment here: NotImplementedError should be used for arguments which are correct but the code is actually not (yet) implemented. However, if the argument (combination) is wrong, it's more typical to raise ValueError. I have a feeling here it's the second case, but I let you judge.

Reply:
@tepals in the following check 
`elif len(self._zlevels) < 1 and focus_points is not None: raise NotImplementedError("If focus_points are used initial zlevels must also be provided")`
Why is it required? As even if z levels are not provided and focus points are provided, the following piece of code in _tileacq.py takes care
`    def _get_zstack_levels(self, focus_value):
        """
        Calculate the zstack levels from the current focus position and zsteps value
        :returns (list(float) or None) zstack levels for zstack acquisition.
          return None if only one zstep is requested.
        """
        # When initial z level is one or None, use the current focus value
        if len(self._init_zlevels) <= 1:
            return [focus_value, ]`
 Following up question, in autofocus_roi.py, I do not see where does the code depends on the manual focus value before autofocus begins. Could you guide me if I need to incorporate that or is it already incorporated?

Reply:
You are correct, so raising the error is not necessary. 

autofocus_roi.py takes the current position of the focus component as the starting position, it might be useful to clarify this in the autofocus_roi docstring.